README.md
# toYaml - A json to yaml serializer  #

toYaml is a very simple and superfast JSON to YAML serializer - encoder - dumper - whatever you like ;)  
NEW: from 1.0.0 you can use toYAML directly from the command line

## Features and Support ##

Javascript Objects and Arrays are supported by extending Object.prototype and Array.prototype.
Right now toYaml features cicular dependency resolution and relational links.


## YAML Spec ##

right now it doen't support the spec completely. It just works for my use case ;) But feel free to pull it and extend it to become a full YAML serializer.

## Technology ##

toYaml substantially is a deep copier! It works in a two pass process! First it does some preprocessing, where it searches for linked objects and enumerates them in the hashes linkHashIn (the links under the key references), and linkHashOut (the referenced object under the key of the linking object). It thereby also creates an enumerated new Object which it passes to the YAML parser. The enumerated object contains each object under the key 'obj' and a unique id for reference under the key "id". The second stage is a pass over the newly constructed Object and outputs the objects with proper indentation and references.

## Usage from Command line ##

Usage is simple:

    npm install toYaml
    
then from the command line 

a) File mode (pass in a well formatted JSON file! check the [Online JSON parser](http://json.parser.online.fr/)  for well formatted JSON)   

    toYaml filename.js [-o outputname](@see --help to get more options)   
      
b) HERDOC or pipe mode (like cat file >> toYaml or toYAM << HEREDOC>>)  will outpu to STDOUT       
  
    toYaml <<HEREDOC
    {"hello":"world"}
    HEREDOC

c) pass JSON directly as first command line argument only single line (for multiline usse b))
    toYaml '{"hello":"world"}'                    
    
## Usage as a module##

Usage is simple:

    npm install toYaml

    YAML=require ('toYaml');

    Then you can use it in an OO fashion style:
    calling YAML.extendObjects(true);
    {'YAML': {'aint':{'markup': 'language'}}}.toYaml();

    ot if you orefer the functional style:
    YAML.toYaml({'YAML': {'aint':{'markup': 'language'}}});

## Test cases ##

call node test.js to run the tests 
or call toYaml yamayam.js to transform it into a yaml file 
 
## Options ##
* you can enable links creation by passing {enableLinks:true} to toYaml({'enableLinks':true})
Default is on; But some serializers might not understand it, so you can turn it off, by passing false;
* if you need to reencode with yaml from TJ Holowaychuk you can use the option {yamlCompatible:true} to make the output look even more human readable on lists, default is false and will fail to reencode into JSON

* there is a security mechanism for circularity which defaults to a depth level of 50 iterations. If you want to set deeper limits set maxLevel to the depth you need.

* usePadding: uses a look ahead algorithm to see which is the longest hashtag in a group and pads out the others to match the maximum length  

* setDefaults: set default options to be used by all calls to YAML: e.g. YAML.setDefaults({YAMLCompatible:true});

## Exported functions ##

- toYaml(obj, opts) or obj.toYaml(opts) ... the YAML serializer
- setDefaults(defaults) ... a function for setting the default options

//mainly for testing purpose, returns the output of preprocessing links
- preProcessLinks (obj, opts) or obj.preProcessLinks(opts)

## Dependencies ##

optimist

## Todos ##

* extend and test yaml 1.3 spec cases;)
* make the second parsing phase even faster by just traversing over the history created in the first pass (including the indent level in the first pass and saving it under a key in the new Object)

# License 

(The MIT License)

Copyright (c) 2010 Gregor Schwab <isimpl@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
1.0.0 / 2011-06-09
==================
  * added support for using toYaml from the command line   

0.0.9 / 2011-06-09
==================
  * package.json parsing fix
     
0.0.7 / 2010-12-30
==================

  * now works with real world objects
  * support for functional links and functional objects (although tis will not reencode into TJ's yaml parser yet)
  * heavy refactoring and testing ;)

0.0.7 / 2010-12-30
==================

  * Changed name to toYaml
  
0.0.6 / 2010-12-30
==================

  * Change the maxDepth Limit back to 50, cause there was a bug with an infinite loop
  
0.0.5 / 2010-12-30
==================

  * Change the maxDepth Limit to 1000, cause its merely an infinite loop protection
  
0.0.5 / 2010-12-30
==================

  * Added options for making jsonToYaml not extend Array.prototype by default
  
0.0.3 - 0.0.4 / 2010-12-30
==================

   * Added circular dependency checks and some more test cases
  
0.0.2 / 2010-12-29
==================

  * Added options for making jsonToYaml compatible by default to TJ Holowaychuk's <tj@vision-media.ca> yaml
  
0.0.1 / 2010-12-29
==================

  * Initial release


test.js
YAML=require('./toYaml');
_=require('./support/underscore/underscore');
var sys=require('sys');

YAML.setDefaults({yamlCompatible:true});


preProcessLinks=YAML.preProcessLinks;

//a test for seeing if preprocessing works
var link=[{s:"hello"}];
var refactoredObj = preProcessLinks([{x:link},{y:7, link:link, assafdsf:link}]);

if((refactoredObj.obj[0].obj.x.obj[0].obj.s.obj)=="\'hello\'") 
  console.log("preProcessLinks new Object composition Test passed");
else 
  console.log("ERROR: preProcessLinks new Object composition Test did not pass"); 


//Test preprocessing on a circular object
var helper = {'test':{'threeway':{'circularity':'circularObject2'}}}
var circularObject1={'circ': { 'name': helper.test, 'id': 123 }}
var circularObject2={'someAttr': { 'attr': circularObject1.circ, 'id': 456 }}
helper.test.threeway.circularity=circularObject2;


try {
  var refactoredObj =  preProcessLinks(helper, {'maxLevel':50});
  if(refactoredObj.obj.test.obj.threeway.obj.circularity.obj.someAttr.obj.attr.obj.name.link=='object1') 
    console.log("preProcessLinks on circular Object Test passed\n");
  else throw new Error()
}catch (e){
   console.log("ERROR: preProcessLinks on circular Object did not pass\n". e); 
   return;
}

//console.log('*********preprocess\n', sys.inspect(refactoredObj, false, 15));

var link={'subsub':"x"};
var complexObj=[{'anObject':{'asubObj':link}}, {x:"abc", y:{x: link, y:[{as:"sff", f:34},{as:"sfdf", f:35},{as:"sasff", f:37}]}},{f:link}];
var testObj= { greg: { name: 'greg', tel: 123 }};
//see the ouput of some objects Just visually


console.log('\n\n------------a circular object with enableLinks: true -------------');
yaml_text = YAML.toYaml(helper, {enableLinks:true});
console.log (yaml_text);
//testCases.add('a circular Object', circularObject1);
console.log('------------OO Style-------------');
//OO Style
YAML.extendObjects(true);
var yaml_text = testObj.toYaml();
YAML.extendObjects(false);
console.log (yaml_text);
console.log('\n\n------------functional Style-------------');
//functional Style
yaml_text = YAML.toYaml(testObj);
console.log (yaml_text);
//test enableLinks option
console.log('\n\n------------functional Style with enableLinks: true-------------');
yaml_text = YAML.toYaml(complexObj, {enableLinks:true});
console.log (yaml_text);
console.log('\n\n------------functional Style with enableLinks: false-------------');
yaml_text = YAML.toYaml(complexObj, {enableLinks:false});
console.log (yaml_text);
console.log('\n\n------------a function reference-------------');
var f= function(){};
yaml_text = YAML.toYaml({a_function:f, some:{other:[{linkToAFunction:f}]},and:{someMoreFunc:function(){}}});
console.log (yaml_text);
//encode back to JSON (please npm install yaml to do this or git update --init --recursive submodules)



var yaml=YAML.toYaml({});
if (YAML.toYaml({})==="") {
  console.log('Test "'+  "an empty object" + '" PASS'+ '\n'
  +'------------------------------------------------------\n'); 
}
else {
  throw new Error('Test didn\'t pass: '+ "an empty object"+ '\n'
  +'------------------------------------------------------\n');
}
  
if (YAML.toYaml([])==="") 
  console.log('Test "'+  "an empty array" + '" PASS'+ '\n'
  +'------------------------------------------------------\n'); 
else throw new Error('Test didn\'t pass: '+ "an empty object"+ '\n'
  +'------------------------------------------------------\n');

/*
 * Here is where the real tests begin 
 */
YAMLDecode=require('./support/yaml/lib/yaml');_=require('./support/underscore/underscore');
var testCases=[]
testCases.add = function(desc, obj){
  testCases.push({'desc':desc, 'obj':obj});
}
testCases.add('hash in hash', testObj);
testCases.add('a complex Object containing 2 links', [{'anObject':{'asubObj':link}}, {x:"abc", y:{x: link, y:[{as:"sff", f:34},{as:"sfdf", f:35},{as:"sasff", f:37}]}},{f:link}]);
testCases.add('list in list', [[1,2,3,[4,5]]]);
testCases.add('a simple string', "hello greg");



console.log('\n ****RUNNING TESTS *****\n');
for (var i=0; i<testCases.length;i++){
  var testCase=testCases[i];
  var yaml=YAML.toYaml(testCase.obj, {enableLinks:false});
  var json=YAMLDecode.eval(yaml);
  if (!_.isEqual(json, testCase.obj)) throw new Error('Test didn\'t pass: '+ testCase.desc+ '\n'
  +'------------------------------------------------------\n'
  +yaml);
  else console.log('Test "'+ testCase.desc + '" PASS'+ '\n'
  +'------------------------------------------------------\n'
  +yaml);
}
console.log('all tests passed - Thank you for running tests');
/*var json=YAMLDecode.eval(yaml_text);
console.log(json)
//deep test fo equal using _
_=require('./support/underscore/underscore');
console.log(_.isEqual(json, testObj));*/

cl.js
#!/usr/bin/env node
/*
 * toYaml command line
 *
 * MIT License (@see License)
 * @author        Gregor Schwab
 * @copyright     (c) 2011 Gregor Schwab
 * Usage Command Line: ./toYaml yamyam.js (@see Readme.md)
 * @requires optimist
 */      
var YAML=require ('./toYaml');        
var logging=false; //set to true to turn on logging
var log=function(){if(logging)console.log('toYaml \n' + Array.prototype.slice.call(arguments) + '\n')}
var sys = require('util');
var constants = require('constants');
var path=require('path'), fs=require('fs');
var command=process.ARGV[1];
var lastPath=path.basename(command);
var argv = require('optimist').usage('Usage: $0 filenameA [filenameB ... -o fileNameOutA -o fileNameOutB] or \n $0 \'JSON\' (don\'t forget to include in \'\' wrappers!)').argv;
var cl=true; //command line  
//console.log(sys.inspect((/.*nmd/).test(argv['$0']), false, 20));
if (!((/.*toYaml/).test(argv['$0']))) {throw Error();} //if not called from commandline

//argv -o normalization
//if there is only one argument inr argv.o optimist doesn't create an array
if(typeof argv.o == 'string') var tmp = argv.o; argv.o=[];argv.o[0]=tmp;
//console.log(sys.inspect(argv, false, 20));           

if (argv.help || argv._[0]=="/?"){console.log('Usage: '+argv.$0+' filenameA [filenameB ... -o fileNameOutA -o fileNameOutB] \n or ' +argv.$0+ ' \'JSON\' (don\'t forget to include it in \'\' wrappers!)'); return;}     
               
var toYaml=function toYamlFn(fileArr, outArr, callback){
    //console.log(sys.inspect(fileArr, false, 20), sys.inspect(outArr, false, 20));
    // Relative or absolute path
    function respondError(msg, context) {
      if (msg.errno===constants.ENOENT) {console.log('ERROR: no such file \"'+ msg.path + '\" perhaps you wanted to use option -o');}
      else console.log("Error: "+sys.inspect(msg));
    }    

    fileArr.forEach(function(fileName, index, fileArr){ 
      try {        
        //input = path.join(__dirname, path.basename(fileName));
        input=fileName;
        log("parsing: "+input); 
        fs.stat(input, function (err, stats) {
          if (err) {
              return respondError(err);
          }
          log("stats: "+sys.inspect(stats)); 
          fs.open(input, 'r', stats.mode, function (err, fd) {
            if (err) return;
            fs.read(fd, stats.size, 0, "utf8", function (e, data) {
              if (e) {
                return respondError(e)
              }
              log("data: "+ data);    
              jsonObj = JSON.parse(data);
              var yaml = YAML.toYaml(jsonObj);
              log("yaml: "+ yaml);              
              handleResponse(yaml);
            });
          });//end fs.open                       
          function handleResponse(yaml){
            writeToFile(yaml)
          }

          function writeToFile(yaml){//write the file to disk
            var buf=new Buffer(yaml, encoding='utf8');
            var outputName = outArr[index]; //if ther is only one argument inr argv.o optimist doesn't create an array
            if( outputName ){outputName=outputName.replace(/([^\.]+)$/,'yaml')}
            else {
              outputName=input.replace(/([^\.]+)$/,'yaml')              
            }
            log(argv.o instanceof Array, sys.inspect(argv), outputName)
            fs.open(outputName, 'w', stats.mode, function (e, fd) {
              fs.write(fd, buf, 0, buf.length, 0, function(err, written){
                if (err) {
                  respondError(err)
                }
                fs.close(fd, function(err){
                  if (err) {
                    respondError(err)
                  }                       
                })
              })
            })        
          }
        })//end fs.stat
      }catch(e){
        respondError(e);
      }//end try..catch        
    })//end forEach
  }//end toYaml   
                                                        
//see if we were passed no argument (means read from stdin)
if((argv._.length))  {
  //try to see if we were passed a file as first parameter  
  try{
    fs.statSync(argv._[0])        
    toYaml(argv._, argv.o);   
  }
  catch(e){                 
       log(argv._[0])   
       //test if we were passed a JSON object as first argument we just output the parsed Yaml file 
       var jsObj = JSON.parse(argv._[0])   
       console.log(YAML.toYaml(jsObj))  
  }         
}      
else{
  process.stdin.resume();
  process.stdin.setEncoding('utf8');
  var data="";
  process.stdin.on('data', function (chunk) {
    data +=chunk; 
  });

  process.stdin.on('end', function () {     
    console.log(YAML.toYaml(JSON.parse(data))) 
  });
}


toYaml.js
// toYAML - Core - Copyright Gregor Schwab <isimpl@gmail.com> (MIT Licensed)
var sys = require('sys');
var defaults = {
    enableLinks: true,
    //default
    yamlCompatible: false,
    //default
    usePadding: true,
    extendObjects: false,
    maxLevel: 100,
    showHidden:false,
    maxLevelMessage: function() {
        return 'Max Object Depth has been reached: maxLevel ' + this.maxLevel + ' please set maxLevel to a higher level if your object is deeper than this level.'
    }
}

function setDefaults(def) {
   helper.extend(defaults, def);
}

function extendObjects(bool) {
    if (bool) {
        Object.prototype.toYaml = OOToYaml;
        Array.prototype.toYaml = OOToYaml;
    } else {
        delete Object.prototype.toYaml;
        delete Array.prototype.toYaml;
    }
}

function toYaml(obj, options){
  return yamlSerializer(options).toYaml(obj);
}

function preProcessLinks(obj, options){
  return yamlSerializer(options).preProcessLinks(obj);
}

exports.extendObjects = extendObjects;
exports.toYaml = toYaml
exports.preProcessLinks = preProcessLinks
exports.setDefaults = setDefaults

function OOToYaml(options) {
    return exports.toYaml(this, options);
}

function yamlSerializer(options) {
    var o,
    obja = [],
    newoba = [],
    Path = [],
    oPath = [];
    var opts = {};
    var idCounter = 0;
    var maxLevelError = false;
    var maxLevelKeyPath = "";
    
    helper.extend(opts, defaults, options);
    
    function toYaml(obj) {
        o = obj;

        var newObj = _preProcessLinks(obj);

        if (maxLevelError) {
            console.log(opts.maxLevelMessage() + "\nKeypath: " + maxLevelKeyPath)
        }
        var header = "";
        var ret = walkToYaml(newObj);
        if (ret !== "") return header + ret + "\n"
        else return ""
    }

    function walkToYaml(obj, l, before) {
        if (typeof before === 'undefined') var before = "";
        if (typeof l === 'undefined') var l = 0;
        //if (typeof obj === 'undefined'){throw new Error('obj undefinend context: before '+before)}
        var level = l;
        var out = "";
        if (opts.enableLinks) {
            var ret = searchLink(obj);
            //lookup the id of our object in the linksIn linksOut hashes
            out += ret.text;
            if (ret.stop === true) {
                //if it's a link node, we have to stop iteration here
                return out;
            }
        }

        if (Object.prototype.toString.call(obj.obj) === '[object Array]') {
            var i = 0,
            len = obj.obj.length;
            for (; i < len; i++) {
                //for each thang in the array
                out += '\n';
                //draw a new line
                out += drawIndent(level, 'array');
                //draw the indent
                out += walkToYaml(obj.obj[i], level + 1, 'array');
                //draw the inner thang
            }
            return out;
        }
        if (typeof obj.obj === 'object') {
            var i,
            index = 0;
            if (opts.usePadding) {
                //preprocess the object to find the perfect indent
                var maxKeyLength = 0;
                for (i in obj.obj) {
                    if (!obj.obj.hasOwnProperty(i)) continue;
                    maxKeyLength = Math.max(i.length, maxKeyLength);
                }
            }

            for (i in obj.obj) {
                if (!obj.obj.hasOwnProperty(i)) return out;
                if (index === 0 && !opts.yamlCompatible) {
                    //draw no newline on first item if we don't need to be yaml compatible
                    } else {
                    out += '\n';
                    //draw a new line
                    out += drawIndent(level, 'object');
                }

                out += i + ": ";
                //draw the attribute name with padding
                if (opts.usePadding) {
                    var padding = maxKeyLength - i.length;
                    out += drawIndent(padding);
                }
                out += walkToYaml(obj.obj[i], level + 1, 'object');
                //draw the inner thang
                index++;
            }
            //console.log('obj level ', level, 'out ', out);
            return out;
        }
        //if(typeof obj === 'string') return '"'+obj+'"'
        out+=obj.obj;
        return out;
    }

    function drawIndent(level, thang) {
        var indent = "";
        if (typeof thang == 'undefined') {
            //level variable is a padding instead
            var pad = level;
            for (var j = 0; j < pad; j++) {
                indent += " ";
            }
            //identation
            return indent;
        }
        //console.log("drawIndent ", level, thang)
        if (thang === "object") {
            for (var j = 0; j < level; j++) {
                indent += "  ";
            }
            //identation
        } else if (thang === "array") {
            for (var j = 0; j < level; j++) {
                indent += "  ";
            }
            //identation with a -
            indent += "- ";
        }
        return indent;
    }

    function preProcessLinks(obj) {
        o = obj;
        return _preProcessLinks(obj);
    }


    function _preProcessLinks(obj, l, k, kp) {
        var level = (typeof l === 'undefined') ? 0: l;
        if (typeof k === 'undefined')
        k = '',
        kp = ''
        if (kp !== '')
        kp = kp + '.' + k
        else
        kp = k;

        var newObj = {};
        newObj.level = level,
        newObj.obj = obj;
        newObj.key = k
        newObj.keyPath = kp;
        
        //the maximum Depth of Iterations may not be exceeded
        if (level > opts.maxLevel) {
            maxLevelError = true;
            maxLevelKeyPath = i + "." + maxLevelKeyPath;
            newObj.obj = opts.yamlCompatible ? "" : "[MAX LEVEL REACHED]"
            return newObj;
        }

        // Primitive types cannot have properties
        switch (typeof obj) {
        case 'undefined':
            newObj.obj = opts.yamlCompatible ? "" : "UNDEFINED";
            return newObj;

        case 'string':
            newObj.obj = JSON.stringify(obj).replace(/'/g, "\\'")
            .replace(/\\"/g, '"')
            .replace(/(^"|"$)/g, "'");
            return newObj;

        case 'number':
        case 'boolean':
            newObj.obj = obj;
            return newObj;

        }
        // For some reason typeof null is "object", so special case here.
        if (obj === null) {
            newObj.obj = "null";
            return newObj;
        }

        // Look up the keys of the object.
        var visible_keys = Object.keys(obj);
        var keys = opts.showHidden ? Object.getOwnPropertyNames(obj) : visible_keys;

        // check for empty objects
        /*if (helper.isEmpty(obj)) {
            newObj.obj = '';
            return newObj;
        }*/

        // Dates without properties can be shortcutted
        if (helper.isDate(obj)) {
            //&& keys.length === 0) {
            newObj.obj = obj.toUTCString();
            return newObj;
        }
        
        //its an object or function
        //generate a local uId
        var uId = uniqueId('object');
        //each object gets an id also if its linked
        newObj.id = uId;
        
        //check if its a circular link first
        //CHECK HERE FOR CIRCLES LEAVE THIS HERE ! OTHERWISE IT WILL PUSH PRIMITIVE OBJECTS ON THE CIRCLE PATH, YOU DON'T WANT THAT!
        var idx;
        if ((idx = checkCircle(newObj)) > -1) {
            return newObj;
        }
        
        //check in history if its a link if yes return the obj
        //history of onjects and keeps linksIn and links Out of each object
        //is link
        if ((idx = history(newObj)) > -1) {
            return newObj;
        }

        // all Functions get shortcutted right now
        if (typeof obj === 'function' ){//&& keys.length === 0) {
          if (helper.isRegExp(obj)) {
            newObj.obj = opts.yamlCompatible ? "" : '[regexp]';
            newObj.type='function'
            return newObj;
          } else {
            var name = obj.name ? ': ' + obj.name : '';
            newObj.obj = opts.yamlCompatible ? "" : '[Function' + name + ']';
            newObj.type='function'            
            return newObj;            
          }
        }

        if (Object.prototype.toString.call(obj) === '[object Array]') {
            newObj.obj = [];
            var i = 0,
            len = obj.length;
            //newObj.obj = [];
            for (; i < len; i++) {
                //for each thang in the array
                newObj.obj.push(_preProcessLinks(obj[i], level + 1, i, kp));
                //copy on without change
            }
            removeFromCirclePath(newObj);
            return newObj;
        }
        if (typeof obj === 'object') {
            newObj.obj = {};
            var i,
            index = 0;
            for (i in obj) {
                if (!obj.hasOwnProperty(i)) continue;
                //newObj.obj[i]={};
                //newObj.obj[i].key=i;
                newObj.obj[i] = _preProcessLinks(obj[i], level + 1, i, kp);
                //draw the inner thang
                index++;
            }
            //pop the object from path before returning
            removeFromCirclePath(newObj);
            return newObj;
        }
    }

    function searchLink(newObj) {
        //id
        if (newObj.ref) return {
            'text': '&' + newObj.id,
            'stop': false
        };
        //a reference
        if (newObj.link) return {
            'text': '*' + newObj.link,
            'stop': true
        };
        //a link
        return {
            'text': "",
            'stop': false
        }
    }

    function checkCircle(newObj) {
        var idx;
        if ((idx = oPath.indexOf(newObj.obj)) > -1) {
            var refObj = Path[idx];
            //if (typeof refObj.obj !== "function") {
                //for now don't support functions
                refObj.ref = true;
                if (typeof refObj.refObjects == "undefined") refObj.refObjects = [];
                if (typeof refObj.circleObjects == "undefined") refObj.circleObjects = [];
                refObj.refObjects.push(newObj);
                refObj.circleObjects.push(newObj);
                newObj.link = refObj.id;
                newObj.obj = refObj.obj;
            //}
            return idx;
            //if its a circle dont push on path
        }
        Path.push(newObj);
        //"push" object to the Path
        oPath.push(newObj.obj);
        return - 1;
    }
    function removeFromCirclePath(obj) {
        Path.pop();
        oPath.pop();
        return obj;
    }

    function history(newObj) {

        //search for links in the history
        var idx = -1;
        if ((idx = obja.indexOf(newObj.obj)) > -1) {
            //loop
            var key = newoba[idx].id;
            //hash lookup
            var refObj = newoba[idx];
            //hash lookup
            //if (typeof refObj.obj !== "function") {
                //for now don't support functions
                refObj.ref = true;
                if (typeof refObj.refObjects == "undefined") refObj.refObjects = [];

                refObj.refObjects.push(newObj);
                //linkHashOut[newObj.id]=key;//this dock is for new objects link to first object = the links
                newObj.link = key;
                newObj.obj = refObj.obj;
            //}
            return idx;
            //if its a link dont push on history
        }
        //create a new history object
        obja.push(newObj.obj);
        newoba.push(newObj);
        return idx;
    }
    // Generate a unique integer id (unique within the entire client session).
    // Useful for temporary DOM ids.

    function uniqueId(prefix) {
      var id = idCounter++;
      return prefix ? prefix + id : id;
    }    
    //closure functions
    return {
        'toYaml': toYaml,
        'preProcessLinks': preProcessLinks
    }
}
//end serializer

var helper={}

helper.isArray = function isArray(ar) {
  return ar instanceof Array ||
         Array.isArray(ar) ||
         (ar && ar !== Object.prototype && isArray(ar.__proto__));
}


helper.isRegExp = function isRegExp(re) {
  var s = '' + re;
  return re instanceof RegExp || // easy case
         // duck-type for context-switching evalcx case
         typeof(re) === 'function' &&
         re.constructor.name === 'RegExp' &&
         re.compile &&
         re.test &&
         re.exec &&
         s.match(/^\/.*\/[gim]{0,3}$/);
}


 helper.isDate = function isDate(d) {
  if (d instanceof Date) return true;
  if (typeof d !== 'object') return false;
  var properties = Date.prototype && Object.getOwnPropertyNames(Date.prototype);
  var proto = d.__proto__ && Object.getOwnPropertyNames(d.__proto__);
  return JSON.stringify(proto) === JSON.stringify(properties);
}

 helper.isString = function isString(obj) {
  return !!(obj === '' || (obj && obj.charCodeAt && obj.substr));
}

// Is a given object empty?
helper.isEmpty = function isEmpty(obj) {
if (helper.isArray(obj) || helper.isString(obj)) return obj.length === 0;
 for (var key in obj) if (hasOwnProperty.call(obj, key)) return false;
 return true;
};

// Extend a given object with all the properties in passed-in object(s).
helper.extend = function extend(obj) {
  Array.prototype.slice.call(arguments, 1).forEach(function(source) {
    for (var prop in source) obj[prop] = source[prop];
  });
  return obj;
}




yaml.js

var sys=require('sys');

var enableLinks=true; //default
var objh={}, links=[];
var idCounter = 0;


Object.prototype.toYaml=OOToYaml;
Array.prototype.toYaml=OOToYaml;

exports.toYaml= toYaml;

var link={'subsub':"x"};
yaml_text = ([{'anObject':{'asubObj':link}}, {x:"abc", y:{x: link, y:[{as:"sff", f:34},{as:"sfdf", f:35},{as:"sasff", f:37}]}},{f:link}]).toYaml();
console.log (yaml_text);

function OOToYaml(options){
  return toYaml(this, options);
}

function toYaml(obj, options){
  if (typeof options.enableLinks!==undefined)
    enableLinks=options.enableLinks;
  if (enableLinks) preProcessLinks(obj);
  console.log(links);
  return walkToYaml(obj);
}
function walkToYaml(obj, l, before) {
    if (typeof before === 'undefined') var before=""; 
    if (typeof l === 'undefined') var l=0;
    var out = "";
    var level=l;
    if (Object.prototype.toString.call(obj) === '[object Array]') {
        var i = 0, len = obj.length;
        for ( ; i < len; i++ ) { //for each thang in the array

            out += '\n'; //draw a new line             
            out += drawIndent(level, 'array'); //draw the indent


          out += walkToYaml(obj[i], level+1, 'array'); //draw the inner thang
        }
        return out;
    }
    if (typeof obj === 'object') {
        //if (enableLinks ){
        //  var match= history(obj);
        //}
        var i, index=0;
            
        //preprocess the object to find the perfect indent
        var maxKeyLength=0;
        for ( i in obj ){
          maxKeyLength=Math.max(i.length, maxKeyLength);
        }
        if (enableLinks){
          var ret=searchLink(obj);
          out += ret.text;
          if (ret.stop===true){
            return out;
          }
        }
        for ( i in obj ) {
          if (!obj.hasOwnProperty(i)) return out;
          //out+="level " + level +   ' index ' + index + ' before ' + before; 
          if (index!==0 || before!=="array"){  //if its the first object after an array don't draw newline nor draw an indent, else we use the arrays indentation
          //else {
            out += '\n'; //draw a new line             
            out += drawIndent(level, 'object');
          }
          //console.log("maxKeyLength ", maxKeyLength)
          var padding=maxKeyLength - i.length+2;       
          out += i+": "; //draw the attribute name with padding
          out += drawIndent(padding);
          out += walkToYaml(obj[i], level+1, 'object') ; //draw the inner thang
          index++;
        }
        return out;
    }
    return obj;
}


function drawIndent(level, thang){
  var indent="";  
  if(typeof thang== 'undefined'){ //level variable is a padding instead
    var pad=level;
    for (var j=0; j<pad; j++) {indent += " ";}//identation  
    return indent;    
  }
  //console.log("drawIndent ", level, thang)

  if (thang === "object"){
    for (var j=0; j<level; j++) {indent += "  ";}//identation  
  }else if (thang === "array"){
    for (var j=0; j<level; j++) {indent += "  ";}//identation with a -  
    indent +="- ";
  }
  return indent; 
}

// Generate a unique integer id (unique within the entire client session).
// Useful for temporary DOM ids.

function uniqueId(prefix) {
  var id = idCounter++;
  return prefix ? prefix + id : id;
}

function preProcessLinks(obj){
    var out = "";
    if (Object.prototype.toString.call(obj) === '[object Array]') {
        var i = 0, len = obj.length;
        for ( ; i < len; i++ ) { //for each thang in the array
          out += preProcessLinks(obj[i]); 
        }
        return out;
    }
    if (typeof obj === 'object') {
        history(obj);
        var i, index=0;
        for ( i in obj ) {
          out += preProcessLinks(obj[i]) ; //draw the inner thang
          index++;
        }
        return out;
    }
}

function searchLink(obj){
  var len= links.length;
  for (var j=0; j<len; j++) {
    var item=links[j];
    if(item.linkedTo===obj && typeof item.touched==='undefined'){
      console.log('searchLink match item ', sys.inspect(item), obj);
      item.touched=true;
      var retVal={'text':'&'+item.linkedToId, 'stop':false}
      return retVal;
    }else if(item.linkedTo===obj && typeof item.touched!=='undefined'){
      var retVal={'text':'*'+item.linkedToId, 'stop':true}
      return retVal;  
    }
  }
  return {'text':"", 'stop':false}
}
function history(obj){

  //search the history
  var match = null;
  for (key in objh){
    var value=objh[key];
    if(obj==value){
      console.log('match item ' , key, sys.inspect(value), obj);
      links.push({'object': obj, 'linkedTo':value, 'linkedToId':key});
    }
  } 
  if (match==null){
    //create a new history object
    var uId=uniqueId('object');
    objh[uId]=obj;
    obj=uId;
    match={'uId':uId, 'first':true}
  }//return object itsself
  
  return match; 
}
